### DISCUSSION TITLE
Fully Implicit Mesh Scheme

### DISCUSSION THREAD

**Discussion by TLWise** [libMesh/libmesh#3812] [Q&A]
Thank you for any help you may be able to provide with this. Is it possible without code modification to implement a fully implicit mesh scheme using libmesh combined with Petsc? The details of which is present here:

Problem Definition:
Given a typical 2nd order Poisson problem. 
     Au = f, 

Find u in V such that a(u,rho) = (f,rho) for all rho in V. 
Approximate the problem using a sequence of finite dimensional subspaces V_h is a proper subset of V, where the discrete problem reads as,
     Find u_h in V_h such that a_h(u_h,rho) = (f_h,rho) for all rho in V_h.
    
     A_hU_h = F_h 

The continuous problem is defined with the boundary conditions, B(u) = g on the surface of the domain cap_gamma. The boundary conditions can be a combination of function values and/or partial derivatives. of u on the boundaries.

Initially Matrix A_h (and also U_h andF_h) have not incorporated any BCs, corresponding to natural BCs.
S_h(cap_gamma) denotes as degrees of freedom of the vector U_h that are related to the boundary cap_gamma.

Fully implicit method:
1) Do not modify A_h, which is a Neumann Matrix due to the natural BCs.
2) Prescribe the Dirichlet values components of the vectors U_h and F_h, belonging to S_h(cap_gamma), before and after each iteration step.
All components of the solution vector have to be treated as unknowns by all components of the iterative solvers.

**Comment by roystgnr** [comment]
By "each iteration step" here, do you mean each Krylov step?

If so, you might need to ask the PETSc people.  I don't see anything in [their manual pages](https://petsc.org/release/manualpages/KSP/) that would serve as a `KSPSetCallbackToModifySolutionAtEachStep`, and something like that is what you'd need to do this right.  You could hack something together where you set a maximum of 1 Krylov iteration (and enable preconditioner reuse) in their solver, and then do the Dirichlet prescription in between each of these 1-step "solves" (along with your own stopping criteria).  That would be sufficient for academic experiments, though I'd never want to see anything like it in a production code.
